### PR TITLE
Validate exchange

### PR DESCRIPTION
**Pull request by rhlsthrm**

### CODE REVIEW

File: modules/node/src/appRegistry/appRegistry.service.ts [hunk 1]
```diff
@@ -1,39 +1,103 @@
...
   constructor(
     private readonly nodeService: NodeService,
+    private readonly configService: ConfigService,
+    private readonly swapRateService: SwapRateService,
     private readonly appRegistryRepository: AppRegistryRepository,
   ) {}
 
-  private appProposalMatchesRegistry(proposal: AppInstanceInfo, registry: AppRegistry): boolean {
+  private async getValidSwaps(): Promise<AllowedSwaps> {
+    const allowedSwaps: AllowedSwaps = [
```
Comment (LayneHaber) [inline modules/node/src/appRegistry/appRegistry.service.ts]: will we need to add in a swap to go back to eth from dai?

Comment (rhlsthrm) [reply modules/node/src/appRegistry/appRegistry.service.ts]: Yep, we sure will. We aren't even calculating and broadcasting the rates, so doesn't make sense to do yet.

File: modules/node/src/appRegistry/appRegistry.service.ts [hunk 2]
```diff
@@ -1,39 +1,103 @@
...
+      throw new Error(
+        `Swap from ${params.initiatorDepositTokenAddress} to ` +
+          `${params.responderDepositTokenAddress} is not valid. Valid swaps: ${JSON.stringify(
+            validSwaps,
+          )}`,
+      );
+    }
+
+    // |our rate - derived rate| / our rate = discrepancy
+    const derivedRate = parseEther(
```
Comment (LayneHaber) [inline modules/node/src/appRegistry/appRegistry.service.ts]: if were doing eth --> dai and dai --> eth allowed well need to account for the `assetId` of both parties here

File: modules/node/src/appRegistry/appRegistry.service.ts [hunk 3]
```diff
@@ -1,39 +1,103 @@
...
+    logger.log(
+      `Derived rate is ${derivedRate.toString()}, within ${ALLOWED_DISCREPANCY_PCT}% ` +
+        `of our rate ${ourRate.toString()}`,
+    );
   }
 
-  private validateLinkedTransfer(params: NodeTypes.ProposeInstallParams): void {
-    if (toBig(params.responderDeposit).gt(Zero)) {
+  private async validateLinkedTransfer(params: NodeTypes.ProposeInstallParams): Promise<void> {
+    if (bigNumberify(params.responderDeposit).gt(Zero)) {
```
Comment (LayneHaber) [inline modules/node/src/appRegistry/appRegistry.service.ts]: should check if its negative as well, maybe just check that its exactly 0

Comment (rhlsthrm) [reply modules/node/src/appRegistry/appRegistry.service.ts]: Can you do a negative deposit? I think checking 0 makes sense.

**Pull request merged**